pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -1377,11 +1351,19 @@ func (c *PodAffinityChecker) satisfiesExistingPodsAntiAffinity(pod *v1.Pod, meta
 	}
 
 	// Iterate over topology pairs to get any of the pods being affected by
-	// the scheduled pod anti-affinity rules
+	// the scheduled pod anti-affinity rules where all its anti-affnity rules
+	// are matched by the incoming pod
+	podToMatchingTermsCount := make(map[string]int)
 	for topologyKey, topologyValue := range node.Labels {
-		if topologyMaps.topologyPairToPods[topologyPair{key: topologyKey, value: topologyValue}] != nil {
-			glog.V(10).Infof("Cannot schedule pod %+v onto node %v", podName(pod), node.Name)
-			return ErrExistingPodsAntiAffinityRulesNotMatch, nil
+		podSet := topologyMaps.topologyPairToPods[topologyPair{key: topologyKey, value: topologyValue}]
```

Comment:
Why do we need to do this? The whole point of precalculating topology pairs is to avoid looking at individual pods on nodes. This change removes a big part of that optimization.

Reply:
I am probably missing something here. Case 2 in your comment should be handled correctly without checking individual pods. In your example, the incoming pod matches matches anti-affinity of the existing pod on nodeA. So, we would put "region:r1" and "zone:z1" in the matching topology pair map. Since nodeA and nodeB have the "region:r1" label, none of these nodes are considered feasible for the incoming pod. This is the right behavior.

Reply:
I'm confused now... Shouldn't we check both labelSelector and topologyKey, and **only if both match**, it's marked as a match?

According to your explanation, it sounds like a (1) full match on labelSelector and (2) partial match on topologyKey.

Reply:
Hmm... I see your point. I went back and read our documentation again. There are two main points from the docs that I think can resolve the issue:
1. The rules are of the form “this pod should (or, in the case of anti-affinity, should not) run in an X if that X is already running one or more pods that meet rule Y”.
1. All matchExpressions associated with requiredDuringSchedulingIgnoredDuringExecution affinity and anti-affinity must be satisfied for the pod to be scheduled onto a node.

Based on the above rules, anti-affinity and affinity should be processed differently. Here what I think is the algorithm to address the requirements:


1. **Existing pod anti-affinity**: If the incoming pod matches the label selector and namespace of any anti-affinity term of an existing pod, the topologyKey and the corresponding label value of the node should be added to the topology pair map. The reason is that this term would be violated if the incoming pod is scheduled on in that topology. So, here we don't need to check all terms. Any term that is matched, should be added. (Our implementation before this PR was valid)

1. **Incoming pod anti-affinity**: If any existing pod on a node matches any terms of the incoming Pod anti-affinity, the topology key and its corresponding value on the node should be added to the map of topology pairs. This is similar to the above rule.

1. **Incoming pod affinity**: This is where things are different. For an affinity rule to be considered satisfied, all the terms of the affinity rule must match. Based on a previous decision, a single pod (as opposed to a group of pods) must match all the terms for the affinity rule to be considered satisfied. I don't have a great idea on how we can use our topology pair map for checking pod affinity efficiently. I think that the implementation that we had in #62211 would work better for  matching affinity, but I might be wrong.

Sorry if this is not consistent with my previous comments on this PR. I hope my understanding is correct this time.

Reply:
Hmm...  It seems that:
- for Affinity, we're always good at the understanding that all terms must be matched - everyone agrees on this
- but for AntiAffinity, it sounds like we revert the understanding of AntiAffinity from "all" to "any"... which is denied in my previous PR #68173

If so, I'd like to **re-evaluate** the comment in closed PR: https://github.com/kubernetes/kubernetes/pull/68173#issuecomment-418484896 to see if we agree or not with that right now.

Where we came to a consensus on that:

```
- we have an existing pod on a node:
    - labels: {"b": ""}
    - PodAntiAffinity: term {"b" exists}
- then we got an incoming pod:
    - labels: {"a": ""}
    - PodAntiAffinity: term {"a" exists} and term {"b" exists}
```

In this case, incoming pod should be approved to land on the node.

And also the symmetric case should be of same result: incoming pod can be scheduled to the node:

```
- we have an existing pod:
    - labels: {"a": ""}
    - PodAntiAffinity: term {"a" exists} and term {"b" exists}
- then we got an incoming pod:
    - labels: {"b": ""}
    - PodAntiAffinity: term {"b" exists}
```

PS: think of the env as an one-node cluster, and simply suppose topologyKey is the same value.

Reply:
Maybe we can start a poll to dev/users to see in terms of AntiAffinity, which option makes more sense lol..

Reply:
> but for AntiAffinity, it sounds like we revert the understanding of AntiAffinity from "all" to "any"... which is denied in my previous PR #68173

@Huang-Wei That's right. My comment was incorrect for the anti-affinity case.

In both of the cases in your previous example, the pod should NOT be scheduled to the node (assuming that the topoloygKey is "hostname"). In your first example, the second term of the anti-affinity of the incoming pod matches the existing pod. So, it cannot land on the node. In the second example, the second term of the anti-affinity of the existing pod matches and again the pod cannot be scheduled.

Reply:
> In both of the cases in your previous example, the pod should NOT be scheduled to the node (assuming that the topoloygKey is "hostname").

I agree (and also agreed) on this :) It was also my first impression on this issue.

Then to proceed delivering a fix, which way do you prefer:

- open a new PR to combine reasonable fixes from both old and this PR
- reopen old PR and incorporate/merge reasonable fixes from this PR
- keep this PR and incorporate reasonable/merge fixes from old PR

Reply:
@Huang-Wei I don't have any strong preference. As long as the implementation is correct, I am fine with any PR. Since reopening the old PR would take less time, we can do that. I think we should also ensure that affinity is processed correctly.

Reply:
Got it. Thanks @bsalamat.